pre_commit_dbt/utils.py

```diff
@@ -415,3 +419,109 @@ def __call__(
                 result[key] = value
 
         setattr(namespace, self.dest, result)
+
+
+def find_file_with_suffix(filename, suffix, root_path):
+    for file in glob.glob(f"{str(root_path)}/**/{filename + '.' + suffix}"):
+        return Path(file)
+
+
+def clean_yml_path(yml_path):
+    double_slash_index = yml_path.find("//")
+    return yml_path[double_slash_index + 2 :]
+
+
+def get_related_sqls(
+    yml_path,
+    paths: Sequence[str],
+    nodes: Dict[str, Any],
+    include_ephemeral: bool = False,
+):
+    yml_path = Path(yml_path)
+    yml_path_parts = yml_path.parts
+
+    root_path = yml_path_parts[0]
+    dbt_patch_path = "/".join(yml_path_parts)
+
+    for key, node in nodes.items():
+        if (
+            not include_ephemeral
+            and node.get("config", {}).get("materialized") == "ephemeral"
+        ):
+            continue
+
+        if node["patch_path"] and dbt_patch_path in node["patch_path"]:
+            if ".sql" in node["original_file_path"]:
+                target_sql_name = f"{root_path}/{node['original_file_path']}"
+                if target_sql_name not in paths:
+                    paths.append(target_sql_name)
+
+    return paths
+
+
+def get_related_yml(
+    sql_path,
+    paths: Sequence[str],
+    nodes: Dict[str, Any],
+    include_ephemeral: bool = False,
+):
+    for key, node in nodes.items():
+        if (
+            not include_ephemeral
+            and node.get("config", {}).get("materialized") == "ephemeral"
+        ):
+            continue
+
+        if node["path"] and node["path"] in sql_path:
+            root_folder = Path(node["root_path"]).name
+
+            # Original patch_path has 'project\\path\to\yml.yml'
+            patch_path = Path(node["patch_path"])
+            # Remove the project_name from patch_path
+            clean_patch_path = patch_path.relative_to(*patch_path.parts[:2])
+
+            target_yml_path = f"{root_folder}/{clean_patch_path}"
+            if target_yml_path not in paths:
+                paths.append(target_yml_path)
+
+    return paths
+
+
+def get_missing_file_paths(
+    paths: Sequence[str],
+    manifest: Dict[str, Any] = None,
+    include_ephemeral: bool = False,
+):
+    nodes = manifest.get("nodes", {})
+
+    if nodes:
+        for path in paths:
+            suffix = Path(path).suffix.lower()
+            if suffix == ".sql":
+                paths = get_related_yml(path, paths, nodes, include_ephemeral)
```

Comment:
You're oeverriding path on every iteration of the loop, I didn't test this but looks like a bug.

Reply:
For what I was able to test, it didn't result in any bug. But it can certainly end up in a bug as the list can grow and results re-used in future loop cases.

Newest commit will use `paths` only for looping, and a `paths_with_missing` that contains the original `*paths` and those appended on the "search" methods